Feature API実装完了

toriyamaa opened the pull request:
APIの実装完了しました。
レビューをよろしくお願いいたします。

toriyamaa committed:
invoice-api　新規作成

toriyamaa committed:
controllerクラス、Resourceクラス、Serviceクラス新規作成

toriyamaa committed:
invoice-dao新規作成

toriyamaa committed:
新たにプロジェクト作成しなおしたので古いプロジェクト削除

toriyamaa committed:
dao層切り出しのために
プロジェクト作成しなおし

toriyamaa committed:
検索、ID指定取得、登録機能を一通り動作するように実装

toriyamaa committed:
データ取得処理修正
　検索条件追加

toriyamaa committed:
エラーハンドリング実装中

toriyamaa committed:
エラーハンドリング、入力チェック実装

toriyamaa committed:
エラーハンドリング実装完了

toriyamaa committed:
不要なコメントアウト削除

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/common/constant/MessageType.java:
```diff
@@ -0,0 +1,41 @@
+	 *
+	 * @return the code
+	 */
+	// メソッド
```
？

toriyamaa replied:
不要コメント削除しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/common/validation/DateTimeCompareValidator.java:
```diff
@@ -0,0 +1,104 @@
+				return true;
+			}
+
+			if (endDateObj.compareTo(startDateObj) >= 0) {
```
LocalDateクラスには、以下のメソッドがあるのでそちらを使用したほうがいいと思います。
・isAfter
・isBefore

toriyamaa replied:
isBeforeをメソッド使うように修正しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/common/validation/DateTimeFormatValidator.java:
```diff
@@ -0,0 +1,53 @@
+	@Override
+	public boolean isValid(String value, ConstraintValidatorContext context) {
+
+		if (value == null) {
```
このバリデーターはnullを許容するで合ってますか？
もう少し、コメントを書くようにしてください。

toriyamaa replied:
このバリデーターはnullを許容するで合ってます。
コメント記述しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/common/validation/DateTimeCompareValidator.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * The Class DateTimeCompareValidator.
+ */
+public class DateTimeCompareValidator implements ConstraintValidator<DateTimeCompare, Object> {
```
バリデータのjavadocは記述するようにお願いします。
何を許容するか、挙動が不明確のため後で混乱を招きます。

toriyamaa replied:
javadoc記述しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/common/validation/DateTimeCompareValidator.java:
```diff
@@ -0,0 +1,104 @@
+			String end = (String) beanWrapper.getPropertyValue(endDay);
+
+			if (start == null || end == null) {
+				return true;
```
値がnullな時は許容するで問題ないのか不明。
getPropertyValueでプロパティが取得できなかった場合、例外が発生しますが問題ないですか？

toriyamaa replied:
前後チェックだけ行うので、nullは許容する想定 でいます。
getPropertyValueでプロパティが取得できなかった場合、例外が発生しますが問題ないですか？
⇒現状のAPIの実装ですとシステムエラーとしてハンドリングされます。このクラス内でcatchしたほうがよろしいでしょうか？

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/controller/InvoiceApiController.java:
```diff
@@ -0,0 +1,88 @@
+ */
+@RestController
+@RequestMapping("api/invoices")
+@Validated
```
クラスに@Validatedをつける意味がありますか？

toriyamaa replied:
getInvoiceメソッドの @Numeric のエラーをハンドリングするためにつけています。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/controller/exception/ApiExceptionHandler.java:
```diff
@@ -0,0 +1,210 @@
+	 * @param message the message
+	 * @return the api error
+	 */
+	private ApiError createApiError(String code, String message) {
```
このメソッド必要でしょうか？

toriyamaa replied:
createApiErrorメソッド削除しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/common/validation/DateTimeCompareValidator.java:
```diff
@@ -0,0 +1,104 @@
+			Matcher mchStart = ptn.matcher(start);
+			Matcher mchEnd = ptn.matcher(end);
+
+			LocalDate startDateObj;
```
初期化すること。

toriyamaa replied:
初期化するように修正しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/common/ApplicationException.java:
```diff
@@ -0,0 +1,87 @@
+	 * @return the field
+	 */
+	@Getter
+	public class Detail implements Serializable {
```
インナークラスを使用している意図はありますか？

toriyamaa replied:
ApplicationExceptionDetailクラスを新たに作成してインナークラス削除しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/config/AppConfig.java:
```diff
@@ -0,0 +1,27 @@
+ * The Class AppConfig.
+ */
+@Configuration
+@PropertySource("classpath:application-api.properties")
```
使用されていないように見えますが必要ですか？

toriyamaa replied:
ApiProperties.javaに @PropertySource を移動しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/controller/exception/ApiExceptionHandler.java:
```diff
@@ -0,0 +1,210 @@
+	 * @see org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler#handleExceptionInternal(java.lang.Exception, java.lang.Object, org.springframework.http.HttpHeaders, org.springframework.http.HttpStatus, org.springframework.web.context.request.WebRequest)
+	 */
+	@Override
+	protected ResponseEntity<Object> handleExceptionInternal(
```
handleExceptionInternalを拡張した実装でも、問題ありませんが
以下のサイトのように、Exception毎に書いたほうが読みやすいと思います。

http://javatechnology.net/spring/spring_exceptionhandler/

toriyamaa replied:
すみません。いったんhandleExceptionInternalを拡張した実装のままでもよろしいでしょうか？
今後実装する場合はException毎に書くようにいたします。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/resource/InvoiceDetailResource.java:
```diff
@@ -0,0 +1,59 @@
+import lombok.Setter;
+
+@Getter
+@Setter
```
@Dataでもいいと思います。

toriyamaa replied:
@Dataに変更しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/resource/InvoiceDetailResource.java:
```diff
@@ -0,0 +1,59 @@
+
+@Getter
+@Setter
+public class InvoiceDetailResource implements Serializable {
```
Javadocを書く癖をつけてください。

toriyamaa replied:
Javadoc記述しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/resource/query/InvoiceListResourceQuery.java:
```diff
@@ -0,0 +1,38 @@
+@Setter
+@DateTimeCompare(message = "InvalidRelation", endDay = "invoiceDateMax", startDay = "invoiceDateMin")
+public class InvoiceListResourceQuery implements Serializable {
+
```
パッケージ名、クラス名がQueryとなっていますが
使用用途を考えると、Dtoでしょうか？

今回修正は不要ですが、Queryという名前のクラスがBeanであるとひと目でわかりづらいため
現場では意味のあるわかりやすい名前をつけるように心がけてください。

toriyamaa replied:
クラスの意味合いとしてはリクエストパラメータを受け取るbeanになります。
今後は意味のわかりやすいクラス名にいたします。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/service/InvoiceService.java:
```diff
@@ -0,0 +1,403 @@
+	public InvoiceDetailResource findInvoice(String invoiceNo) throws ApplicationException {
+
+		//エラーチェック
+		errCheckFindInvoice(invoiceNo);
```
下の請求書データ取得処理と合わせて2度取得SQLが実行されることになりますが
不要ですので処理の見直しをお願いします。

toriyamaa replied:
errCheckFindInvoiceメソッド削除して、findInvoiceメソッド内で請求書データ取得チェックを行うように修正しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/service/InvoiceService.java:
```diff
@@ -0,0 +1,403 @@
+				simpleDateFormat.format(invoice.getUpdateDatetime()));
+
+		//請求書に紐づく注文実績データ取得
+		List<Order> orderList = orderRepository.findByClientNoAndInvoiceTerm(client.getClientNo(),
```
client.getClientNo()も改行しませんか？

toriyamaa replied:
改行しました

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/service/InvoiceService.java:
```diff
@@ -0,0 +1,403 @@
+
+		Invoice invoice = invoiceRepository.findByInvoiceNo(Integer.parseInt(invoiceNo), DelFlg.NOT_DELETE.getValue());
+		if (invoice == null) {
+			ae.addDetailList(MessageDetailType.INVOICE_NOT_EXIST.getDetailCode(), new String[] {}, "");
```
API設計書にないコードを使用しているように見えます。

toriyamaa replied:
API設計書にエラー詳細コードを追加しました。

uenos commented on invoice-api/src/main/java/tigers/cave/webm/invoice/api/service/InvoiceService.java:
```diff
@@ -0,0 +1,403 @@
+		InvoiceRegistrationResultResource invoiceRegistrationResultResource = new InvoiceRegistrationResultResource();
+		invoiceRegistrationResultResource.setInvoiceNo(String.valueOf(invoice.getInvoiceNo()));
+
+		URI resourceUri = MvcUriComponentsBuilder.relativeTo(uriBuilder)
```
この処理は、複数メソッドで出てくるので共通化すべきだと思います。

toriyamaa replied:
URI 構築処理共通化しました。

uenos reviewed (changes requested):
レビュー完了

toriyamaa committed:
レビュー指摘事項修正
>「インナークラスを使用している意図はありますか？」
⇒ApplicationExceptionDetailクラスを新たに作成しました。

toriyamaa committed:
レビュー指摘事項修正
不要なコメント削除しました。

toriyamaa committed:
レビュー指摘事項修正

>バリデータのjavadocは記述するようにお願いします。何を許容するか、挙動が不明確のため後で混乱を招きます。
⇒javadoc記述しました。

>値がnullな時は許容するで問題ないのか不明。
⇒前後チェックだけ行うので、nullは許容する想定

>初期化すること。
⇒LocalDate初期化しました。

>LocalDateクラスには、以下のメソッドがあるのでそちらを使用したほうがいいと思います。
>isAfter
>isBefore
⇒isBeforeメソッドを使用するように修正

toriyamaa committed:
レビュー指摘事項修正
>このバリデーターはnullを許容するで合ってますか？もう少し、コメントを書くようにしてください。
⇒必須かどうかはここではチェックしないのでnullは許容の想定です。コメント記述しました。

toriyamaa committed:
レビュー指摘事項修正
>使用されていないように見えますが必要ですか？
⇒@PropertySource を ApiProperties.javaに移動しました。

toriyamaa committed:
レビュー指摘事項反映
・createApiErrorメソッド削除
・

toriyamaa committed:
レビュー指摘事項修正
・@Getter、@Setterを@Dataに変更
・Javadoc記述

toriyamaa committed:
レビュー指摘事項修正
・2度取得SQLが発行されないように修正
・client.getClientNo()部分改行
・URI構築処理共通化

toriyamaa committed:
エラー詳細コード追加しました。
InvoiceDuplication
OrderNotExist
InvoiceNotExist

toriyamaa committed:
レビュー指摘事項修正
インナークラス削除

toriyamaa committed:
pom.xml修正　findbugs設定

toriyamaa committed:
pom.xml修正

toriyamaa committed:
pom.xml修正

toriyamaa committed:
pom.xml修正

toriyamaa committed:
pom.xml
InvoiceServiceTest.java 作成中

toriyamaa committed:
pom.xml修正

toriyamaa committed:
pom.xml修正

toriyamaa committed:
pom.xml修正

toriyamaa committed:
カバレッジのplugin変更
cobertura ⇒ JaCoCo

toriyamaa committed:
カバレッジツール調整中

toriyamaa committed:
pom.xml jacoco設定追加

toriyamaa committed:
pom.xml jacoco version設定修正

toriyamaa committed:
InvoiceServiceのUT完了
　InvoiceServiceの日付フォーマット修正

toriyamaa committed:
checkstyle対応

toriyamaa committed:
不要なコメントアウト削除

toriyamaa committed:
checkstyle対応

toriyamaa committed:
checkstyleのルールを設定

toriyamaa committed:
checkstyleルール設定

toriyamaa committed:
checkstyleルール修正

toriyamaa committed:
checkstyle設定調整

toriyamaa committed:
checkstyle修正

toriyamaa committed:
google_checks.xml修正

toriyamaa committed:
checkstyleのバージョン変更
invoice-daoのgoogle_checks.xml削除

toriyamaa committed:
google_checks.xml 修正

toriyamaa committed:
google_checks.xml 修正

toriyamaa committed:
pom.xml の checkstyle　設定修正

toriyamaa committed:
以下修正
Integer.valueOf → Integer.parseInt

findbugsの設定修正

toriyamaa committed:
findbugsの除外設定修正

toriyamaa committed:
findbugs-exclude.xmlをinvoice-api、invoice-daoに追加
